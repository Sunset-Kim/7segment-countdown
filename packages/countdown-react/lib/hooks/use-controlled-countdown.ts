import { formatTimeLeft } from "@7segment/core";
import { useState, useEffect, useCallback, useRef } from "react";

/**
 * 제어 가능한 카운트다운 훅을 생성합니다.
 *
 * @param {number} initialSeconds - 카운트다운의 초기 시간(초 단위)
 * @returns {Object} 카운트다운 상태와 제어 함수를 포함하는 객체
 * @property {number} days - 남은 일 수
 * @property {number} hours - 남은 시간
 * @property {number} minutes - 남은 분
 * @property {number} seconds - 남은 초
 * @property {boolean} isRunning - 카운트다운 실행 중 여부
 * @property {Function} start - 카운트다운 시작 함수
 * @property {Function} stop - 카운트다운 중지 함수
 * @property {Function} reset - 카운트다운 초기화 함수
 */
export const useControlledCountdown = (initialSeconds: number) => {
  const [secondsLeft, setSecondsLeft] = useState(initialSeconds);
  const [isRunning, setIsRunning] = useState(false);
  const intervalRef = useRef<NodeJS.Timeout | null>(null);

  const clearCountdownInterval = useCallback(() => {
    if (intervalRef.current) {
      clearInterval(intervalRef.current);
      intervalRef.current = null;
    }
  }, []);

  const stopCountdown = useCallback(() => {
    clearCountdownInterval();
    setIsRunning(false);
  }, [clearCountdownInterval]);

  const tick = useCallback(() => {
    setSecondsLeft((prevSeconds) => {
      if (prevSeconds <= 1) {
        stopCountdown();
        return 0;
      }
      return prevSeconds - 1;
    });
  }, [stopCountdown]);

  useEffect(() => {
    setSecondsLeft(initialSeconds);
  }, [initialSeconds]);

  useEffect(() => {
    if (isRunning) {
      intervalRef.current = setInterval(tick, 1000);
    } else {
      clearCountdownInterval();
    }

    return clearCountdownInterval;
  }, [isRunning, tick, clearCountdownInterval]);

  const start = useCallback(() => {
    if (!isRunning && secondsLeft > 0) {
      setIsRunning(true);
    }
  }, [isRunning, secondsLeft]);

  const stop = useCallback(() => {
    if (isRunning) {
      setIsRunning(false);
    }
  }, [isRunning]);

  const reset = useCallback(() => {
    stopCountdown();
    setSecondsLeft(initialSeconds);
  }, [initialSeconds, stopCountdown]);

  return {
    ...formatTimeLeft(secondsLeft),
    isRunning,
    start,
    stop,
    reset,
  };
};
