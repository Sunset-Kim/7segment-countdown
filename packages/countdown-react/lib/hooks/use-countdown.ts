import { calculateTimeLeft, formatTimeLeft } from "../../../core/dist";
import { useEffect, useState } from "react";

interface CountdownHookOption {
  onEnd?: () => void;
}

/**
 * 주어진 목표 날짜까지의 카운트다운을 관리하는 커스텀 React 훅입니다.
 *
 * @param {Date} targetDate - 카운트다운의 목표 날짜입니다.
 * @param {Object} [option] - 선택적 설정 객체입니다.
 * @param {Function} [option.onEnd] - 카운트다운이 0에 도달했을 때 호출될 콜백 함수입니다.
 *
 * @returns {Object} 남은 시간을 포맷팅한 객체를 반환합니다.
 * @returns {number} returns.days - 남은 일수
 * @returns {number} returns.hours - 남은 시간
 * @returns {number} returns.minutes - 남은 분
 * @returns {number} returns.seconds - 남은 초
 *
 * @example
 * const { days, hours, minutes, seconds } = useCountdown(new Date('2025-01-01'), {
 *   onEnd: () => console.log('카운트다운 종료!')
 * });
 */
export const useCountdown = (
  targetDate: Date,
  option?: CountdownHookOption
) => {
  const [timeLeft, setTimeLeft] = useState(
    calculateTimeLeft({ targetDate, currentDate: new Date() })
  );

  useEffect(() => {
    const interval = setInterval(() => {
      const newTimeLeft = calculateTimeLeft({
        targetDate,
        currentDate: new Date(),
      });
      setTimeLeft(newTimeLeft);

      if (newTimeLeft <= 0) {
        clearInterval(interval);
        option?.onEnd?.();
      }
    }, 1000);

    return () => clearInterval(interval);
  }, [targetDate, option]);

  return formatTimeLeft(timeLeft);
};
