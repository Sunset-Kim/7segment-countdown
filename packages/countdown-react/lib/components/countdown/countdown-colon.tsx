import { useCountdownStyle } from "../../hooks/use-countdown-style-context";
import { Colon, ColonProps } from "../colon";

export const CountDownColon = ({ ...props }: ColonProps) => {
  const { size, color, colonGap: spacing } = useCountdownStyle();
  return <Colon size={size} color={color} colonGap={spacing} {...props} />;
};
