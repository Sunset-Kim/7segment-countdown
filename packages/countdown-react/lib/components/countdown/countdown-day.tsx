import { useCountdownContext } from "../../hooks/use-countdown-context";
import { useCountdownStyle } from "../../hooks/use-countdown-style-context";
import { SevenSegment, SevenSegmentProps } from "../seven-segment";

export const CountdownDay = ({
  ...props
}: Omit<SevenSegmentProps, "digit">) => {
  const { days } = useCountdownContext();
  const daysString = days.toString().padStart(2, "0");
  const { ...styles } = useCountdownStyle();
  return (
    <>
      {daysString.split("").map((day, index) => {
        return (
          <SevenSegment
            key={index}
            digit={parseInt(day)}
            {...styles}
            {...props}
          />
        );
      })}
    </>
  );
};
