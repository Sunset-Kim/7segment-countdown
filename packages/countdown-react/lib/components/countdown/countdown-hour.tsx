import { useCountdownContext } from "../../hooks/use-countdown-context";
import { useCountdownStyle } from "../../hooks/use-countdown-style-context";
import { SevenSegment, SevenSegmentProps } from "../seven-segment";

export const CountdownHour = ({
  ...props
}: Omit<SevenSegmentProps, "digit">) => {
  const { hours } = useCountdownContext();
  const hoursString = hours.toString().padStart(2, "0");
  const { ...styles } = useCountdownStyle();

  return (
    <>
      {hoursString.split("").map((hour, index) => {
        return (
          <SevenSegment
            key={index}
            digit={parseInt(hour)}
            {...styles}
            {...props}
          />
        );
      })}
    </>
  );
};
