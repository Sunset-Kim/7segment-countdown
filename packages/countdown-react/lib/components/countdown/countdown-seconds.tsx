import { useCountdownContext } from "../../hooks/use-countdown-context";
import { useCountdownStyle } from "../../hooks/use-countdown-style-context";
import { SevenSegment, SevenSegmentProps } from "../seven-segment";

export const CountdownSeconds = ({
  ...props
}: Omit<SevenSegmentProps, "digit">) => {
  const { seconds } = useCountdownContext();
  const secondsString = seconds.toString().padStart(2, "0");
  const { ...styles } = useCountdownStyle();
  return (
    <>
      {secondsString.split("").map((hour, index) => {
        return (
          <SevenSegment
            key={index}
            digit={parseInt(hour)}
            {...styles}
            {...props}
          />
        );
      })}
    </>
  );
};
