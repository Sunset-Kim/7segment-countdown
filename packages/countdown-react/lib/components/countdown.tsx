import { useCountdown } from "../hooks/use-countdown";
import { memo, PropsWithChildren, useMemo } from "react";
import { SevenSegmentProps } from "./seven-segment";
import { Colon, ColonProps } from "./colon";
import { SegmentDisplay } from "./segment-display";

export interface CountdownProps
  extends PropsWithChildren,
    Omit<SevenSegmentProps, "digit">,
    ColonProps {
  targetDate: Date;
  onEnd?: () => void;
}

function CountdownContainer(props: CountdownProps) {
  const {
    targetDate,
    children,
    onEnd,
    animationDuration,
    color,
    offColor,
    onColor,
    colonGap,
    size,
  } = props;

  const _targetDate = useMemo(() => targetDate, [targetDate]);
  const { days, hours, minutes, seconds } = useCountdown(_targetDate, {
    onEnd,
  });

  const segmentDisplayProps = {
    animationDuration,
    offColor,
    onColor,
    size,
  };
  const colonProps = {
    color,
    colonGap,
    size,
  };

  return (
    children ?? (
      <>
        <SegmentDisplay
          value={days}
          place={Math.max(2, days.toString().length)}
          {...segmentDisplayProps}
        />
        <Colon {...colonProps} />
        <SegmentDisplay value={hours} {...segmentDisplayProps} />
        <Colon {...colonProps} />
        <SegmentDisplay value={minutes} {...segmentDisplayProps} />
        <Colon {...colonProps} />
        <SegmentDisplay value={seconds} {...segmentDisplayProps} />
      </>
    )
  );
}

export const Countdown = memo(CountdownContainer);
