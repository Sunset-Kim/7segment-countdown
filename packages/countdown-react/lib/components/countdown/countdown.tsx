import { useCountdown } from "../../hooks/use-countdown";

import { PropsWithChildren } from "react";
import { CountdownDay } from "./countdown-day";
import { CountdownMinutes } from "./countdown-minutes";
import { CountdownHour } from "./countdown-hour";
import { CountdownSeconds } from "./countdown-seconds";
import { CountdownStyleProvider } from "../../contexts/coundown-style.context";
import { CountdownProvider } from "../../contexts/countdown.context";
import { SevenSegmentProps } from "../seven-segment";
import { CountDownColon } from "./countdown-colon";
import { ColonProps } from "../colon";

interface CountdownProps
  extends PropsWithChildren,
    Omit<SevenSegmentProps, "digit">,
    ColonProps {
  targetDate: Date;
}

function CountdownContainer({
  targetDate,
  children,
  ...props
}: CountdownProps) {
  const { days, hours, minutes, seconds } = useCountdown(targetDate);

  return (
    <CountdownProvider
      value={{
        days,
        hours,
        minutes,
        seconds,
      }}
    >
      <CountdownStyleProvider value={{ ...props }}>
        {children ?? (
          <>
            <Countdown.Days />
            <Countdown.Colon />
            <Countdown.Hours />
            <Countdown.Colon />
            <Countdown.Minutes />
            <Countdown.Colon />
            <Countdown.Seconds />
          </>
        )}
      </CountdownStyleProvider>
    </CountdownProvider>
  );
}

export const Countdown = Object.assign(CountdownContainer, {
  Days: CountdownDay,
  Minutes: CountdownMinutes,
  Hours: CountdownHour,
  Seconds: CountdownSeconds,
  Colon: CountDownColon,
});
