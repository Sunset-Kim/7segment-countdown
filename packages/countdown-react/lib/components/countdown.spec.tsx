import { describe, it, expect, vi, beforeEach } from "vitest";
import { render, screen } from "@testing-library/react";
import { Countdown } from "./countdown";
import { useCountdown } from "../hooks/use-countdown";

// useCountdown 훅을 모킹합니다
vi.mock("../hooks/use-countdown", () => ({
  useCountdown: vi.fn(() => ({
    days: 1,
    hours: 2,
    minutes: 3,
    seconds: 4,
  })),
}));

describe("Countdown", () => {
  beforeEach(() => {
    vi.clearAllMocks();
  });

  it("기본 props로 렌더링", () => {
    const targetDate = new Date();
    const { container } = render(<Countdown targetDate={targetDate} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it("자식 컴포넌트가 있을 때 렌더링", () => {
    const targetDate = new Date();
    const { container } = render(
      <Countdown targetDate={targetDate}>
        <div>Custom child</div>
      </Countdown>
    );
    expect(container.firstChild).toMatchSnapshot();
    expect(screen.getByText("Custom child")).toBeInTheDocument();
  });

  it("SegmentDisplay와 Colon에 올바른 props 전달", () => {
    const targetDate = new Date();
    render(
      <Countdown
        targetDate={targetDate}
        color="red"
        size={100}
        colonGap={50}
        offColor="gray"
        onColor="green"
        animationDuration={300}
      />
    );

    const segmentDisplays = screen.getAllByRole("img");
    expect(segmentDisplays[0]).toHaveAttribute("width", "100");

    const colons = screen.getAllByRole("img", { name: "" });
    expect(colons[0]).toHaveAttribute("height", "200");
  });

  it("useCountdown 훅이 올바른 인자로 호출됨", () => {
    const targetDate = new Date();
    const onEnd = vi.fn();
    render(<Countdown targetDate={targetDate} onEnd={onEnd} />);

    expect(vi.mocked(useCountdown)).toHaveBeenCalledWith(targetDate, { onEnd });
  });

  it("날짜가 변경되면 리렌더링", () => {
    const { rerender } = render(<Countdown targetDate={new Date()} />);
    expect(vi.mocked(useCountdown)).toHaveBeenCalledTimes(1);

    rerender(<Countdown targetDate={new Date(Date.now() + 1000)} />);
    expect(vi.mocked(useCountdown)).toHaveBeenCalledTimes(2);
  });
});
