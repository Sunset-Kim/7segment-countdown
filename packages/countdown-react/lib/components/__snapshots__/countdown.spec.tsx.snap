// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Countdown > 기본 props로 렌더링 1`] = `
<svg
  height="200"
  role="img"
  viewBox="0 0 100 200"
  width="100"
>
  <g>
    <polygon
      class="segment on"
      data-testid="segment-A"
      id="segment-A"
      points="10,10 20,2.5 80,2.5
   90,10 80,17.5 20,17.5"
      style="fill: dodgerblue; transition: fill 300ms;"
    />
    <polygon
      class="segment on"
      data-testid="segment-B"
      id="segment-B"
      points="90,10 97.5,20 97.5,80
   90,90 82.5,80 82.5,20"
      style="fill: dodgerblue; transition: fill 300ms;"
    />
    <polygon
      class="segment on"
      data-testid="segment-C"
      id="segment-C"
      points="90,90 97.5,100 97.5,160
   90,170 82.5,160 82.5,100"
      style="fill: dodgerblue; transition: fill 300ms;"
    />
    <polygon
      class="segment on"
      data-testid="segment-D"
      id="segment-D"
      points="10,170 20,162.5 80,162.5
   90,170 80,177.5 20,177.5"
      style="fill: dodgerblue; transition: fill 300ms;"
    />
    <polygon
      class="segment on"
      data-testid="segment-E"
      id="segment-E"
      points="10,90 17.5,100 17.5,160
   10,170 2.5,160 2.5,100"
      style="fill: dodgerblue; transition: fill 300ms;"
    />
    <polygon
      class="segment on"
      data-testid="segment-F"
      id="segment-F"
      points="10,10 17.5,20 17.5,80
   10,90 2.5,80 2.5,20"
      style="fill: dodgerblue; transition: fill 300ms;"
    />
    <polygon
      class="segment "
      data-testid="segment-G"
      id="segment-G"
      points="10,90 20,82.5 80,82.5
   90,90 80,97.5 20,97.5"
      style="fill: aliceblue; transition: fill 300ms;"
    />
  </g>
</svg>
`;

exports[`Countdown > 자식 컴포넌트가 있을 때 렌더링 1`] = `
<div>
  Custom child
</div>
`;
